jmespath-core/src/main/java/io/burt/jmespath/function/JoinFunction.java

```diff
@@ -0,0 +1,59 @@
+package io.burt.jmespath.function;
+
+import java.util.List;
+import java.util.ArrayList;
+
+import io.burt.jmespath.Adapter;
+
+public class JoinFunction extends JmesPathFunction {
+  public JoinFunction() {
+    super(2, 2);
+  }
+
+  @Override
+  protected <T> T internalCall(Adapter<T> adapter, List<ExpressionOrValue<T>> arguments) {
+    T glue = arguments.get(0).value();
+    T components = arguments.get(1).value();
+    if (isStringArray(adapter, components)) {
+      if (adapter.isString(glue)) {
+        List<T> values = adapter.toList(components);
+        if (values.isEmpty()) {
+          return adapter.createString("");
+        } else {
+          StringBuilder buffer = new StringBuilder();
+          String glueString = adapter.toString(glue);
+          T lastValue = values.get(values.size() - 1);
+          for (T value : values) {
+            buffer.append(adapter.toString(value));
+            if (value != lastValue) {
+              buffer.append(glueString);
+            }
+          }
```

Comment:
Technically, the reference to lastValue could be present multiple times in the array, so it would probably be better to do something like this:
```java
Iterator i = values.iterator();
buffer.append(adapter.toString(i.next()));
while (i.hasNext()) {
  buffer.append(glueString).append(adapter.toString(i.next()));
}
```